[language] Add new FixedPoint32 module to stdlib

bob-wilson opened the pull request:
This adds a fixed-point numeric type with 32 fractional bits. This can be used for things like interest rates, conversion rates, etc. Besides functions to create these fixed-point values, this provides basic truncating multiply and divide functions.

## Motivation

Add support for non-integer values in a fixed-point representation.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added some new tests for this module, including all the interesting corner cases that I could think of.

tzakian commented on language/move-lang/stdlib/modules/fixedpoint32.move:
```diff
@@ -0,0 +1,62 @@
+    // numerator and denominator. This function is for convenience; it is also
+    // perfectly fine to create a fixed-point value by directly specifying the
+    // raw value.
+    public fun create_from_rational(numerator: u64, denominator: u64): T {
```
Do we want to check and error if the `denominator` is zero?

bob-wilson replied:
I could add a check and abort. Is that any better than just letting the VM report an ARITHMETIC_ERROR?

tzakian replied:
I think that we should have an error check at the Move level. It's personal taste, so feel free to ignore, but I always feel weary relying on a VM/execution to bubble up the right error.

bob-wilson replied:
OK, personal taste: My taste is to eventually add a specification block with an aborts_if condition showing that the function is expected to abort when the denominator is zero.

tzakian commented on language/move-lang/tests/functional/fixedpoint32/test.move:
```diff
@@ -0,0 +1,20 @@
+    let f3 = FixedPoint32::create_from_raw_value(FixedPoint32::get_raw_value(copy f2) + 1);
+    let three = FixedPoint32::multiply_u64(9, copy f3);
+    0x0::Transaction::assert(three == 3, three);
+}
```
Not needed now since this isn't being run... but having div-by-zero tests to make sure they error the way they should would be good for both `divide_u64` and `create_from_rational`

tnowacki replied:
Isn't this being run? It's in the functional tests

tnowacki replied:
Probably wan't more edge-case tests like, u64 max used in several places

bob-wilson replied:
Yes, I need to add more tests. It didn't work at all when I tried it yesterday because the functional tests were using the IR version of the stdlib where the new module does not exist.

tnowacki commented on language/move-lang/stdlib/modules/fixedpoint32.move:
```diff
@@ -0,0 +1,62 @@
+	let scaled_denominator = (denominator as u128) << 32;
+	// Divide. This cannot overflow.
+	let quotient = scaled_numerator / scaled_denominator;
+	T { value: (quotient as u64) }
```
Does this cast fail if the number is too large?

dariorussi replied:
it will in the runtime for what is worth

bob-wilson replied:
I don't know what I was thinking when I wrote that comment -- probably that these were only 32 bit values. Of course this can overflow. The current behavior of integer casts is that they truncate, but @vgao1996 is planning to change that. In the meantime, I can add an explicit check here.

dariorussi commented:
how are additions and subtractions performed with this model/implementation?
Also is this PR to have a conversation about the fixed point (which is great) or do we intend to commit this in stdlib?

bob-wilson commented:
> how are additions and subtractions performed with this model/implementation?

We could add functions for those if they are commonly used. In the meantime, you can just add and subtract the raw values. There isn't anything special needed there.

> Also is this PR to have a conversation about the fixed point (which is great) or do we intend to commit this in stdlib?

I'd like to commit it as soon as I can get the tests to work.

tzakian commented:
The switchover has landed! So you should be able to get these working once you get the cycles :)

bob-wilson commented:
> The switchover has landed! So you should be able to get these working once you get the cycles :)

Yes, I got it working today and I'm in the process of writing more tests.

bob-wilson commented:
I've updated this with more tests and comments. I also added an explicit check for underflow in the create_from_rational function, since that seems like it could be a source of unexpected behavior that we can disallow. This is ready for review now.

bob-wilson commented:
One more change: I switched to use Transaction::assert and to report an abort code of 16 for the various arithmetic errors here, after I discovered that the abort code of 10 is generally used for "insufficient balance" errors. After we change the VM to fail on overflow with integer type casts (and perhaps pending various discussions of coding style), I am planning to remove the explicit checks for overflows, which will leave only the one explicit Transaction::assert for underflow in create_from_rational.

bob-wilson commented:
I updated this PR to take advantage of #2805 to report overflows in the type casts.

vgao1996 commented:
I know we probably don't have a use case for now, but do we want to offer something like `multiply_fixed32_fixed32` or `divide_fixed32_fixed32`?

bob-wilson commented:
> I know we probably don't have a use case for now, but do we want to offer something like `multiply_fixed32_fixed32` or `divide_fixed32_fixed32`

I'd rather wait to see what is needed, but that's just my personal preference.

bob-wilson commented:
@bors-libra r=tzakian

bors-libra commented:
:pushpin: Commit 77180b3 has been approved by `tzakian`

<!-- @bors-libra r=tzakian 77180b34fa7abe947ca6325a54ddec8540ef7393 -->

bors-libra commented:
:hourglass: Testing commit 77180b34fa7abe947ca6325a54ddec8540ef7393 with merge cbbd55eb7492c0911f8e918944f48c44de736faa...

bors-libra commented:
:broken_heart: Test failed - [checks-circle_commit_workflow](https://circleci.com/workflow-run/ab962f2a-6dea-4502-8bc1-30856a016f79?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)

bob-wilson commented:
@bors-libra retry

bors-libra commented:
:hourglass: Testing commit 77180b34fa7abe947ca6325a54ddec8540ef7393 with merge 518ad12b17e2871e0b9f6d8f71106113f0debb33...

bors-libra commented:
:sunny: Test successful - [checks-actions_land_blocking_test](https://github.com/libra/libra/runs/531822373), [checks-circle_commit_workflow](https://circleci.com/workflow-run/449fc069-5627-42cd-a913-7efefe8d7dff?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: tzakian
Pushing 518ad12b17e2871e0b9f6d8f71106113f0debb33 to master...
